servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsServiceDiscovererBuilder.java

```diff
@@ -209,75 +202,47 @@ public DefaultDnsServiceDiscovererBuilder ioExecutor(final IoExecutor ioExecutor
-        }
-        return factory.create(new DefaultDnsServiceDiscoverer(
-                ioExecutor == null ? globalExecutionContext().ioExecutor() : ioExecutor, minTTLSeconds, ndots,
-                invalidateHostsOnDnsFailure, optResourceEnabled, queryTimeout, dnsResolverAddressTypes,
-                dnsServerAddressStreamProvider));
+    public ServiceDiscoverer<HostAndPort, InetSocketAddress, ServiceDiscovererEvent<InetSocketAddress>>
+            buildARecordDiscoverer() {
+        return asHostAndPortDiscoverer(build());
     }
 
     /**
-     * Convert this object from {@link String} host names and {@link InetAddress} resolved address to
-     * {@link HostAndPort} to {@link InetSocketAddress}.
+     * Create a new instance of {@link DnsClient}.
      *
-     * @return a resolver which will convert from {@link String} host names and {@link InetAddress} resolved address to
-     * {@link HostAndPort} to {@link InetSocketAddress}.
+     * @return a new instance of {@link DnsClient}.
      */
-    private static ServiceDiscoverer<HostAndPort, InetSocketAddress,
-            ServiceDiscovererEvent<InetSocketAddress>> toHostAndPortDiscoverer(
-            final ServiceDiscoverer<String, InetAddress, ServiceDiscovererEvent<InetAddress>> serviceDiscoverer) {
-        return new ServiceDiscoverer<HostAndPort, InetSocketAddress, ServiceDiscovererEvent<InetSocketAddress>>() {
-            @Override
-            public Completable closeAsync() {
-                return serviceDiscoverer.closeAsync();
-            }
-
-            @Override
-            public Completable closeAsyncGracefully() {
-                return serviceDiscoverer.closeAsyncGracefully();
-            }
-
-            @Override
-            public Completable onClose() {
-                return serviceDiscoverer.onClose();
-            }
-
-            @Override
-            public Publisher<ServiceDiscovererEvent<InetSocketAddress>> discover(final HostAndPort hostAndPort) {
-                return serviceDiscoverer.discover(hostAndPort.hostName()).map(originalEvent ->
-                        new DefaultServiceDiscovererEvent<>(new InetSocketAddress(originalEvent.address(),
-                                hostAndPort.port()), originalEvent.isAvailable())
-                );
-            }
-        };
+    DnsClient build() {
+        DnsClient rawClient = new DefaultDnsClient(
+                ioExecutor == null ? globalExecutionContext().ioExecutor() : ioExecutor, minTTLSeconds, ndots,
+                invalidateHostsOnDnsFailure, optResourceEnabled, queryTimeout, dnsResolverAddressTypes,
+                dnsServerAddressStreamProvider);
+        DnsClientFilterFactory rawFilterFactory = filterFactory;
+        if (applyRetryFilter) {
+            DnsClientFilterFactory retryFilterFactory = new RetryingDnsClientFilter();
+            rawFilterFactory = rawFilterFactory == null ? retryFilterFactory :
+                    retryFilterFactory.append(rawFilterFactory);
```

Comment:
Because we do not expose `DnsClient` as a public API, can we keep existing `ServiceDiscovererFilter` and `ServiceDiscovererFilterFactory` and apply them in `buildSrvDiscoverer` and `buildARecordDiscoverer` methods respectively?
Then we do not need to change the pre-existing retrying filter as we can apply it on top of the returned `ServiceDiscoverer` interface.

Reply:
discussed offline. lets hold off on the `ServiceDiscovererFilterFactory` general APIs for now and re-evaluate after we get more experience with DNS and other implementations if the general APIs are worth it.